File: frontend_flutter_backup/lib/repositories/guides_repository.dart
import '../services/api_service.dart';

class GuideRankingEntry {
  const GuideRankingEntry({
    required this.userId,
    required this.name,
    required this.photo,
    required this.location,
    required this.rating,
    required this.reviewCount,
    required this.rankPoints,
    required this.completedTrips,
  });

  final String userId;
  final String name;
  final String photo;
  final String location;
  final double rating;
  final int reviewCount;
  final int rankPoints;
  final int completedTrips;
}

class GuidesRepository {
  GuidesRepository(this.api);

  final ApiService api;

  Future<List<GuideRankingEntry>> ranking() async {
    final response = await api.dio.get<dynamic>('/guides/ranking');
    return (response.data as List? ?? const []).whereType<Map>().map((raw) {
      final profile = raw.cast<String, dynamic>();
      final user =
          (profile['user'] as Map?)?.cast<String, dynamic>() ?? const {};
      return GuideRankingEntry(
        userId: user['id']?.toString() ?? profile['userId']?.toString() ?? '',
        name: user['name']?.toString() ?? 'Local guide',
        photo: user['avatarUrl']?.toString() ?? '',
        location: '${profile['city'] ?? ''}, ${profile['country'] ?? ''}',
        rating: double.tryParse(profile['rating']?.toString() ?? '') ?? 0,
        reviewCount: profile['reviewCount'] as int? ?? 0,
        rankPoints: profile['rankPoints'] as int? ?? 0,
        completedTrips: profile['completedTrips'] as int? ?? 0,
      );
    }).toList();
  }

  Future<Map<String, dynamic>> mine() async {
    final response = await api.dio.get<dynamic>('/guides/me');
    return (response.data as Map).cast<String, dynamic>();
  }

  Future<void> updateProfile({
    required String name,
    required String country,
    required String city,
    required String bio,
    required List<String> expertise,
    required String? price,
  }) => api.dio.patch<void>(
    '/guides/me',
    data: {
      'name': name,
      'country': country,
      'city': city,
      'bio': bio,
      'expertise': expertise,
      'pricingType': price == null ? 'NONE' : 'HOURLY',
      'price': ?price,
    },
  );

  Future<void> apply({
    required String country,
    required String city,
    required String bio,
    required int experienceYears,
    required List<String> languages,
    required List<String> expertise,
    required List<String> availability,
    required String? price,
    required bool verified,
    required int assessmentScore,
    required String referenceContact,
    required bool codeOfConductAccepted,
  }) async {
    await api.dio.post(
      '/guides/apply',
      data: {
        'country': country,
        'city': city,
        'bio': bio,
        'experienceYears': experienceYears,
        'languages': {
          for (final language in languages)
            language: language == 'Mongolian' ? 'Native' : 'Fluent',
        },
        'expertise': expertise,
        'availability': availability,
        'pricingType': price == null ? 'NONE' : 'HOURLY',
        'price': ?price,
        'verified': verified,
        'assessmentScore': assessmentScore,
        'referenceContact': referenceContact,
        'codeOfConductAccepted': codeOfConductAccepted,
      },
    );
  }
}

File: frontend/src/repositories/guidesRepository.js
import { api } from '../services/api';
import { mapGuide } from '../models';

export const guidesRepository = {
  async all() {
    const { data } = await api.get('/guides');
    return (data || []).map(mapGuide);
  },

  async one(id) {
    const { data } = await api.get(`/guides/${id}`);
    return mapGuide(data);
  },

  async ranking() {
    const { data } = await api.get('/guides/ranking');
    return (data || []).map((raw) => {
      const user = raw.user || {};
      return {
        userId: String(user.id ?? raw.userId ?? ''),
        name: String(user.name ?? 'Local guide'),
        photo: String(user.avatarUrl ?? ''),
        location: [raw.city, raw.country].filter(Boolean).join(', '),
        rating: Number(raw.rating) || 0,
        reviewCount: Number(raw.reviewCount) || 0,
        rankPoints: Number(raw.rankPoints) || 0,
        completedTrips: Number(raw.completedTrips) || 0,
        responseRate: Number(raw.responseRate) || 0,
        acceptanceRate: Number(raw.acceptanceRate) || 0,
        providerCancellationCount: Number(raw.providerCancellationCount) || 0,
        confirmedReportCount: Number(raw.confirmedReportCount) || 0,
        rankingUpdatedAt: raw.rankingUpdatedAt || null,
      };
    });
  },

  async mine() {
    const { data } = await api.get('/guides/me');
    return data;
  },

  async updateProfile(payload) {
    await api.patch('/guides/me', payload);
  },

  async apply(payload) {
    await api.post('/guides/apply', {
      country: payload.country,
      city: payload.city,
      bio: payload.bio,
      experienceYears: payload.experienceYears,
      languages: payload.languages || {},
      expertise: payload.expertise,
      availability: payload.availability,
      pricingType: payload.price == null ? 'NONE' : 'HOURLY',
      price: payload.price,
      referenceContact: payload.referenceContact,
      codeOfConductAccepted: payload.codeOfConductAccepted,
    });
  },

  async applications() {
    const { data } = await api.get('/guides/applications');
    return data || [];
  },

  async reviewApplication(id, payload) {
    const { data } = await api.patch(`/guides/applications/${id}/review`, payload);
    return data;
  },
};
